lib/workers/repository/update/branch/execute-upgrade-commands.ts

```diff
@@ -0,0 +1,199 @@
+import is from '@sindresorhus/is';
+import minimatch from 'minimatch';
+import { mergeChildConfig } from '../../../../config';
+import { GlobalConfig } from '../../../../config/global';
+import type { PostUpgradeTasks } from '../../../../config/types';
+import { addMeta, logger } from '../../../../logger';
+import type { ArtifactError } from '../../../../modules/manager/types';
+import { exec } from '../../../../util/exec';
+import {
+  localPathIsFile,
+  readLocalFile,
+  writeLocalFile,
+} from '../../../../util/fs';
+import { getRepoStatus } from '../../../../util/git';
+import type { FileChange } from '../../../../util/git/types';
+import { regEx } from '../../../../util/regex';
+import { sanitize } from '../../../../util/sanitize';
+import { compile } from '../../../../util/template';
+import type { BranchConfig, BranchUpgradeConfig } from '../../../types';
+
+export interface UpgradeCommandsExecutionResult {
+  updatedArtifacts: FileChange[];
+  artifactErrors: ArtifactError[];
+}
+
+export async function upgradeTaskExecutor(
+  upgradeTask: PostUpgradeTasks | null | undefined,
+  config: BranchConfig,
+  updatedArtifacts: FileChange[],
+  allowedUpgradeCommands: string[] | undefined,
+  allowUpgradeCommandTemplating: boolean | undefined,
+  upgrade: BranchUpgradeConfig,
+  taskType: string
+): Promise<UpgradeCommandsExecutionResult> {
+  let currentUpdatedArtifacts = updatedArtifacts;
+  let artifactErrors: ArtifactError[] = [];
+
+  addMeta({ dep: upgrade.depName });
+  logger.trace(
+    {
+      tasks: upgradeTask,
+      allowedCommands: allowedUpgradeCommands,
+    },
+    `Checking for ${taskType.toLowerCase()} tasks`
+  );
+
+  const commands = upgradeTask?.commands ?? [];
```

Comment:
we should validate allowed commands here, so we can skip file persisting if no allowed commands are available.